File: vigenere.c
#include <cs50.h>
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>

void badInput()
{
    printf("Usage: ./vigenere keyword\n");
    exit(1);
}

int main(int argc, string argv[])
{
    // check keyword
    if (argc != 2)
    {
        badInput();
    }
    for (int i = 1; argv[1][i] != '\0'; i++)
    {
        if (!isalpha(argv[1][i]))
        {
            badInput();
        }
    }
    
    // retrieve keyword
    string keyword = argv[1];
    
    // retrieve keyword length
    int keywordLength;
    for (keywordLength = 0; keyword[keywordLength] != '\0'; keywordLength++);
    
    // retrieve key array
    int keyArray[keywordLength];
    for (int i = 0; i < keywordLength; i++)
    {
        keyArray[i] = islower(keyword[i])
                      ? ((keyword[i] - 'a') % ('z' - 'a' + 1))
                      : ((keyword[i] - 'A') % ('Z' - 'A' + 1));
    }
    
    // retrieve plaintext and store in ciphertext
    string ciphertext = get_string("plaintext:  ");
    
    // encrypt
    for (int i = 0; ciphertext[i] != '\0'; i++)
    {
        if (isalpha(ciphertext[i]))
        {
            ciphertext[i] = islower(ciphertext[i])
                            ? (((ciphertext[i] - 'a' + keyArray[i % keywordLength])
                                % ('z' - 'a' + 1)) + 'a')
                            : (((ciphertext[i] - 'A' + keyArray[i % keywordLength])
                                % ('Z' - 'A' + 1)) + 'A');
        }
    }
    
    // display ciphertext
    printf("ciphertext: %s\n", ciphertext);
    
    return 0;
}

File: vigenere.py
import sys

from cs50 import get_string


def badInput():
    sys.exit("Usage: python vigenere.py keyword")


# check key
if len(sys.argv) != 2:
    badInput()

if not sys.argv[1].isalpha():
    badInput()

# retrieve keyword
keyword = list(sys.argv[1])

# retrieve plaintext and store in ciphertext
ciphertext = list(get_string("plaintext:  "))

keyArray = []
for c in keyword:
    keyArray.append(((ord(c) - ord('A')) % (ord('Z') - ord('A') + 1),
                     (ord(c) - ord('a')) % (ord('z') - ord('a') + 1))
                    [c.islower()])


# encrypt
j = 0
for i in range(len(ciphertext)):
    if ciphertext[i].isalpha():
        ciphertext[i] = chr(
            (((ord(ciphertext[i]) - ord('A') + keyArray[j % len(keyword)]) % (ord('Z') - ord('A') + 1)) + ord('A'),
             ((ord(ciphertext[i]) - ord('a') + keyArray[j % len(keyword)]) % (ord('z') - ord('a') + 1)) + ord('a'))
            [ciphertext[i].islower()])
        j += 1

# display ciphertext
print("ciphertext:", "".join(ciphertext))
